core/src/main/scala/kafka/cluster/Partition.scala

```diff
@@ -49,6 +49,35 @@ import org.apache.kafka.server.log.internals.{AppendOrigin, FetchDataInfo, Fetch
 import scala.collection.{Map, Seq}
 import scala.jdk.CollectionConverters._
 
+/**
+ * Listener receives notification from an Online Partition.
+ *
+ * A listener can be (re-)registered to an Online partition only. The listener
+ * is notified as long as the partition remains Online. When the partition fails
+ * or is deleted, respectively `onFailed` or `onDeleted` are called once. No further
+ * notifications are sent after this point on.
+ *
+ * Note that the callbacks are executed in the thread that triggers the change
+ * AND that locks may be held during their execution. They are meant to be used
+ * as notification mechanism only.
+ */
+trait PartitionListener {
+  /**
+   * Called when the Log increments its high watermark.
+   */
+  def onHighWatermarkUpdated(partition: TopicPartition, offset: Long): Unit = {}
+
+  /**
+   * Called when the Partition has a failure (e.g. goes offline).
+   */
+  def onFailed(partition: TopicPartition): Unit = {}
```

Comment:
It seems in this PR these two do nothing. Are there plans to use them in the future?

Reply:
Taking a closer look, I guess we don't have a Listener that currently does anything in the main code -- so looks like this PR lays the groundwork for one.

Reply:
Yeah, this is needed for [KAFKA-14673](https://issues.apache.org/jira/browse/KAFKA-14673).